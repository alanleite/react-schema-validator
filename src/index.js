export useStateWithValidator from "./useStateWithValidator"
export schema from "./schema-builder"
export validate from "./validate"