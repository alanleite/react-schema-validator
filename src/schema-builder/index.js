import text from './text'
import number from './number'

const builder = {
  text,
  number
}

export default builder