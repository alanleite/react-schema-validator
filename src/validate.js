const validateField = (key, fieldSchema, value) => {
  const fieldErros = []
  fieldSchema.validators.forEach(val => {
    const isValid = val.action(value)
    if (!isValid) {
      fieldErros.push({
        key: key,
        message: val.message
      })
    }
  })
  return fieldErros
}

export default (schema, data) => {
  let errors = []

  // doesn't suport arrays
  if (Array.isArray(data)) return {}

  // for object validations
  if (data && typeof data === 'object') {
    for (const key in schema) {
      errors = [
        ...errors,
        ...validateField(
          key,
          schema[key],
          data[key]
        )
      ]
    }
    // make a better object
    return errors.reduce((obj, current) => {
      if (obj[current.key]) {
        obj[current.key].push(current.message)
        return obj
      } else {
        obj[current.key] = [current.message]
        return obj
      }
    }, {})
  }

  // if it is a raw value
  return validateField(
    '_',
    schema,
    data
  ).map(e => e.message)
}