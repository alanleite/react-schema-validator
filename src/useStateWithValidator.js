import { useState } from 'react'
import validate from './validate'

export default (defaultState, schema) => {
    const [state, setState] = useState(defaultState)
    const setStateValidate = (data) => {
        data._errors = validate(schema, data)
        setState(data)
    }
    return [
        state,
        setStateValidate
    ]
}