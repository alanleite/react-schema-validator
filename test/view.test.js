import React from 'react'
import { render, fireEvent, wait } from 'react-testing-library'
import { useStateWithValidator, schema } from '../src'

const Page = () => {
    const defaultData = { name: 'Jonhy' }
    const definitions = { name: schema.text().max(2) }
    const [state, setState, errors] = useStateWithValidator(defaultData, definitions)

    const changeName = () => {
        setState({
            name: 'Teste'
        })
    }

    return (
        <div>
            {state.name}
            {JSON.stringify(errors)}
            <button onClick={changeName}>Change</button>
        </div>
    )
}

describe('Button component', () => {
    it('deve conseguir renderizar [snapshot]', () => {
        const { container, getByText } = render(<Page />)
        expect(container.firstChild).toMatchSnapshot()
        fireEvent.click(getByText('Change'))
        return wait(() =>
            expect(container.firstChild).toMatchSnapshot()
        )
    })
})