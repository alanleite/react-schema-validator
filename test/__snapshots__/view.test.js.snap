// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component deve conseguir renderizar [snapshot] 1`] = `
<div>
  Jonhy
  <button>
    Change
  </button>
</div>
`;

exports[`Button component deve conseguir renderizar [snapshot] 2`] = `
<div>
  Teste
  {"name":["Este campo não é do tipo texto","Quantidade máxima de caracteres é de 2."]}
  <button>
    Change
  </button>
</div>
`;
